File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s):
        stack = []
        pairs = {
            ')': '(',
            ']': '[',
            '}': '{'
        }

        for b in s:
            if b in "([{":
                stack.append(b)
            else:
                if not stack:
                    return False

                top = stack.pop()

                if pairs[b] != top:
                    return False

        return not stack

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {

        Deque<Character> stack = new ArrayDeque<>();

        for (char b : s.toCharArray()) {

            if (b == '(' || b == '[' || b == '{') {
                stack.push(b);
            } 
            else {

                if (stack.isEmpty()) {
                    return false;
                }

                char top = stack.pop();

                if (b == ')' && top != '(') return false;
                if (b == ']' && top != '[') return false;
                if (b == '}' && top != '{') return false;
            }
        }

        return stack.isEmpty();
    }
}